libsatadrivers/src/ahci.c

```diff
@@ -0,0 +1,956 @@
+/*
+ * Copyright 2019, DornerWorks
+ *
+ * This software may be distributed and modified according to the terms of
+ * the BSD 2-Clause license. Note that NO WARRANTY is provided.
+ * See "LICENSE_BSD2.txt" for details.
+ *
+ * A lot of this code was based on code found at http://wiki.osdev.org/AHCI
+ * which is in the public domain.
+ *
+ * @TAG(DORNERWORKS_BSD)
+ *
+ */
+
+#include <satadrivers/ide.h>
+#include <satadrivers/ahci.h>
+#include <unistd.h>
+#include <string.h>
+#include <sel4/sel4.h>
+#include <platsupport/delay.h>
+
+#include <satadrivers/ahci_types.h>
+#include <utils/zf_log.h>
+
+#define GET_LOWER_32BITS(val) (uint32_t)((uint64_t)(val) & 0xFFFFFFFFLL)
+#define GET_UPPER_32BITS(val) (uint32_t)(((uint64_t)(val) & 0xFFFFFFFF00000000LL) >> 32)
```

Comment:
why not simply `((uint32_t)((val) >> 32))`

Reply:
Same as the `LOWER_32BITS`, added assurance.

Reply:
C rules are pretty clear how things works. This looks a bit like mystification unless there is an explanation why all this mystification this is needed.

Reply:
Right, ok, I'm happy to just follow the suggestion then.